To-do 리스트 Redux를 사용하여 리팩터링 하기

Heejae0704 opened the pull request:
App 컴포넌트에 테스트 작성하고 Redux 적용하였습니다. 
Reducer에 대한 테스트도 작성하였습니다.

Heejae0704 committed:
App 컴포넌트가 리덕스의 store 기반으로 할 일 목록을 보여주는 기능에 대한 테스트

App.jsx에 대한 테스트 코드를 리덕스 기반의 테스트가 되도록 수정하여 Red가 되었습니다.
Green이 되려면 리덕스의 useSelector mock 함수에 넣어주는 Task list가 화면에
랜더링되어야 합니다.

Heejae0704 committed:
App 컴포넌트 리덕스 적용 및 reducer 테스트 작성

App 컴포넌트 테스트를 기반으로 리덕스 적용하고 reducer 테스트 작성

Heejae0704 committed:
Reducer를 별도 파일로 빼고 테스트 파일도 별도로 구성

Heejae0704 committed:
reducer 테스트 커버리지 100% (브랜치)

브랜치 커버리지 100%가 안되는 이유? 함수에 argument를 아예 안 넣는 방식의 테스트를
적용하니 100% 커버됨. 이유는 왜인지 모르겠음...
action.type에서 에러가 나기 때문에, reducer 코드에서 guard clause 적용함

Heejae0704 committed:
테스트 커버리지 100% 달성 및 user interaction 테스트 작성

KJoobin commented on src/App.jsx:
```diff
@@ -1,37 +1,31 @@
-  const { newId, taskTitle, tasks } = state;
+  const { taskTitle, tasks } = useSelector(selector);
 
   function handleChangeTitle(event) {
```
`event` 객체를 받는것이 아닌, value 만 받아서 의존성을 끊어주면 좋을거 같아요!!

Heejae0704 replied:
네, 수정하였습니다.

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,13 +1,50 @@
   ));
 
-  expect(getByText(/추가/)).not.toBeNull();
+  expect(getByText(/아무 것도 하지 않기/)).not.toBeNull();
+  expect(getByText(/그래도 뭐라도 하기/)).not.toBeNull();
```
getBy~~ 와 queryBy~~ 의 차이점에 대해서 아시나요?!

Heejae0704 replied:
해당하는 노드가 없을 때 getBy~는 에러를 던지고, queryBy~는 null을 던진다고 문서에 나와 있네요. 그렇다면 not.toBeNull()이 정상 동작하려면 queryBy~가 맞겠네요!

KJoobin commented on src/reducer.js:
```diff
@@ -0,0 +1,36 @@
+};
+
+export default function reducer(state = initialState, action) {
+  if (!action) return state;
+  switch (action.type) {
```
`switch` 를 사용하지 않는 방법으로 해결해주세요 :)

Heejae0704 replied:
네 커버리지가 100%가 되지 않아서 이것 저것 시도해봤었는데, 다시 guard clause로 바꿔도 커버리지가 100% 되었네요.

KJoobin commented on src/reducer.test.js:
```diff
@@ -0,0 +1,87 @@
+        newId: 100,
+        taskTitle: '',
+        tasks: [],
+      };
+      expect(reducer(initialState, updateTaskTitle('할 일을 하기'))).toEqual({
```
Arrange 와 Assert 사이에는 빈줄이 있으면 가독성이 좋습니다 :)

[AAA pattern](https://github.com/goldbergyoni/javascript-testing-best-practices#-%EF%B8%8F-12-structure-tests-by-the-aaa-pattern)

Heejae0704 replied:
넵!

KJoobin commented on src/reducer.test.js:
```diff
@@ -0,0 +1,87 @@
+      });
+    });
+  });
+  describe('action type is not one of the defined actions', () => {
```
혹시 [context](https://www.npmjs.com/package/jest-plugin-context) 에 대해서 아시나요? :)

Heejae0704 replied:
describe는 하나의 기능을 묶어주는 역할을 하고, context는 그 기능 내에서 특정 상황이나 조건에서 진행하는 테스트를 묶어주는 역할을 하는군요. 예시를 보니 'with ....'라는 컨벤션으로 설명을 달아주니 간결하고 명확한 것 같아서 코드에 적용했습니다.

KJoobin reviewed (commented):
잘 진행해주셨습니다 :)

테스트 코드를 짤때 가독성을 조금만 신경 써 주시면 더 좋을거같아요!!

+ commit 단위를 TDD 단위(Red - Green - Refactoring)로 진행해주시면 좋을거 같습니다 :)

Heejae0704 committed:
리뷰 코멘트 반영 및 테스트 코드 리펙터링 (Refactor)

테스트 코드의 가독성을 높이기 위해 각각의 테스트를 분리하였습니다.
customRender 함수를 이용하여 필요한 query를 빼서 쓰는 방식을 적용했습니다.

KJoobin commented on src/App.jsx:
```diff
@@ -1,43 +1,37 @@
     <Page
       taskTitle={taskTitle}
-      onChangeTitle={handleChangeTitle}
+      onChangeTitle={(e) => handleChangeTitle(e.target.value)}
```
해당 내용을 Page Component 내부에서 진행하는게 의존성을 더 좋을거같은데 어떻게 생각하시나요?

Heejae0704 replied:
그렇다면 아예 Input Component에서 실제 Input tag 내에서 해당 처리를 해주는 것이 좋을 거 같아서 그 방향으로 수정했습니다.

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,14 +1,62 @@
+  });
 
-  expect(getByText(/추가/)).not.toBeNull();
+  it('changes title', () => {
```
저는 event 를 테스트 할때 컨벤션처럼 `listens for ~~ event on { action }` 을 사용합니다 :)

Heejae0704 replied:
꿀팁이네요! :)

KJoobin commented on src/App.test.jsx:
```diff
@@ -1,14 +1,62 @@
 
-  expect(getByText(/추가/)).not.toBeNull();
+  it('changes title', () => {
+    const { getByLabelText } = customRender();
+    fireEvent.change(getByLabelText(/할 일/), { target: { value: 'a' } });
```
빈칸을 넣어주시면 가독성이 더 좋을거 같아요 :)
```suggestion
    const { getByLabelText } = customRender();
    
    fireEvent.change(getByLabelText(/할 일/), { target: { value: 'a' } });
```

Heejae0704 replied:
넵

Heejae0704 committed:
코드 리뷰 코멘트 반영

Heejae0704 committed:
Page 컴포넌트를 Container 컴포넌트들로 리펙터 (REFACTOR)

App 컴포넌트의 리덕스 관련 기능을 두 개의 Container 컴포넌트로 분리하고
Page 컴포넌트는 삭제

KJoobin commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    ));
+  }
+
+  afterEach(() => {
+    jest.clearAllMocks();
+  });
```
👍

KJoobin commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    const { queryByText, getByDisplayValue } = customRender();
+
+    expect(queryByText(/추가/)).not.toBeNull();
+    expect(getByDisplayValue(/New Title/)).not.toBeNull();
```
queryBy 로 수정하면 좋을거 같아요 :)
```suggestion
    expect(queryByDisplayValue(/New Title/)).not.toBeNull();
```

Heejae0704 replied:
넵!

KJoobin commented on src/InputContainer.test.jsx:
```diff
@@ -0,0 +1,47 @@
+    expect(queryByText(/추가/)).not.toBeNull();
+    expect(getByDisplayValue(/New Title/)).not.toBeNull();
+
+    fireEvent.click(queryByText(/추가/));
+
+    expect(dispatch).toBeCalledWith({
+      type: 'addTask',
+    });
```
추가와 New Title 이 있는것과, 추가 버튼을 눌렀을때 addTask dispatch 가 작동하는 테스트를 분리하면 어떨까요? :)

Heejae0704 replied:
네 분리하였습니다!

KJoobin reviewed (commented):
InputContainer 와 ListContainer 를 만드니깐 테스트도 훨씬 깔끔해진거 같아요!!

Heejae0704 committed:
코드 리뷰 코멘트 반영

- InputContainer 테스트 일부 분리
- getBy -> queryBy 수정: null 체크 시에는 반드시 queryBy 사용!

KJoobin reviewed (approved):
회고 남겨주시면 머지하도록 하겠습니다 :)
